Adds HTTP Client Adapter

JasonTheAdams opened the pull request:
This PR will add the Client Adapter so that the PHP AI Client uses the WP functions to fulfill requests.

Commit:
feat: starts client adapter work

felixarntz commented on includes/http/class-wp-ai-client-client-adapter.php:
```diff
@@ -0,0 +1,187 @@
+		$response = \wp_remote_request( $url, $args );
+
+		if ( \is_wp_error( $response ) ) {
+			throw new WP_AI_Client_Network_Exception(
```
Thinking we should probably add more exception classes to `WordPress/php-ai-client` and try using those here too.

This has been in my head for a while, but I haven't raised it since it wasn't urgent. But now that 0.1.0 is out, we should do some polishing. I think a `Request_Exception` or `Network_Exception` would help, or maybe even multiple custom exceptions for different purposes, including project-scoped extensions of `InvalidArgumentException` and `RuntimeException`. We could also have a `AiClientExceptionInterface` that all exception classes implement so that a caller can easily do a "catch-all" for AI specific exceptions during an LLM call.

Obviously this is more of a discussion for the other repo, but seeing this here prompted me to bring it up :)

JasonTheAdams replied:
@felixarntz What would you think about about moving forward with this and then making a subsequent PR that specifically handles exceptions in this package?

felixarntz replied:
@JasonTheAdams Works for me. But then let's simply use `Exception` here and add `TODO` comments instead of introducing our own one - just so that it's clear it's temporary.

JasonTheAdams replied:
Done!

Commit:
Merge branch 'trunk' into http-adapter

dkotter reviewed (commented):
In testing out implementing the PHP AI Client package in one of our plugins, ran into an issue where our E2E tests all fail. The issue for us is we install a test plugin during our E2E tests that hooks into `pre_http_request` and returns mocked responses (so we don't have to hit actual APIs every time we run tests).

This doesn't work when requests are sent by the PHP AI Client package as it doesn't use the WordPress HTTP API. I had thought that this PR might solve that but still getting the same problems while testing this. Apologies if I'm missing something obvious here but are there additional steps required to ensure this adapter is used?

felixarntz commented:
@dkotter You're on the bleeding edge 😆 

This is all early WIP. This adapter is indeed supposed to fix the problem you mentioned. It's not a bug in the PHP AI Client SDK since it is WordPress agnostic, but that kind of thing is exactly why this package exists.

As far as I can tell, this PR implements the adapter but doesn't use it yet, but that will be added. Apologies for the confusion, but it might be a bit _too_ early to test this package in particular as it's not meant to be usable just yet. That'll hopefully be different in a week or so :)

dkotter commented:
> It's not a bug in the PHP AI Client SDK since it is WordPress agnostic, but that kind of thing is exactly why this package exists.

Correct, not calling that out as a bug, just hadn't thought of that until I opened a PR and saw all our tests fail 😭 

> As far as I can tell, this PR implements the adapter but doesn't use it yet, but that will be added. Apologies for the confusion, but it might be a bit too early to test this package in particular as it's not meant to be usable just yet. That'll hopefully be different in a week or so

No worries, sounds good. Realize I'm jumping in early here, was just trying to help do some testing on this PR to help get it merged but realize there's lots of things still in progress across all these work streams

Commit:
Merge branch 'trunk' into http-adapter

Commit:
refactor: rename file to use PSR-4 autoloading

Commit:
chore: adds WordPress stubs to PHPStan

Commit:
fix: corrects type issues

Commit:
chore: cleans up phpstan file

Commit:
feat: adds nyholm/psr7 package for psr17 factory

Commit:
feat: adds HTTPlug discovery strategy

Commit:
feat: adds PSR17 discoverability

JasonTheAdams commented:
@felixarntz This is back to you for review! A couple things to note:

First, I updated the class names and such to reflect PSR-4, and I updated some types stuff to pass PHPStan. We'll simply have more use of `@var` and ignoring in this project to get things to play nicely between PSR and WordPress. It is what it is. 🤷‍♂️ 

Second, I added a `DiscoveryStrategy` implementing class. As far as I can tell, as it's not well documented, this is how to make this discoverable to the HTTPlug system. You'll notice `WP_AI_Client_Discovery_Strategy` has an `init()` method, which is what would be called to prepend this to the discovery system. I believe you said you'd like this to be ready for another system to call said methods instead of this package bootstrapping itself?

Third, I added the `nyholm/psr7` package to provide a PSR17 factory. It seems to be the de facto lightweight solution and is PHP 7.2 compatible. My only concern is this issue (https://github.com/Nyholm/psr7/issues/256) which points out some basic PHP 8.4 deprecation fixes that's been neglected.

What do you think? 😄

felixarntz commented on includes/http/WP_AI_Client_Client_Adapter.php:
```diff
@@ -0,0 +1,192 @@
+ *
+ * @since n.e.x.t
+ */
+class WP_AI_Client_Client_Adapter implements ClientInterface {
```
I would simply call this class something like `WordPress_HTTP_Client`:
* While it adapts the WordPress HTTP functions to work with PSR, itself it's simply a PSR client, also per the interface it implements.
* It sends HTTP requests, so it has nothing to do with AI directly :)

JasonTheAdams replied:
Agreed! Resolved in 3998eb7

felixarntz commented on includes/http/WordPress_HTTP_Client.php:
```diff
@@ -0,0 +1,192 @@
+			// TODO: Update to use PHP AI Client exceptions.
+			throw new \Exception(
+				$response->get_error_message(), // phpcs:ignore WordPress.Security.EscapeOutput.ExceptionNotEscaped
+				$response->get_error_code() ? (int) $response->get_error_code() : 0
```
This doesn't work in 99% of scenarios, WordPress error codes are almost always strings with text.

What I would suggest is:
* Either introduce something like a `WP_Error_Exception` that allows to maintain the WordPress error codes too, via an additional property.
* Or rely on one of the HTTP request/response specific exceptions that are being added as part of https://github.com/WordPress/php-ai-client/pull/78.

JasonTheAdams replied:
Let's see if we can get the exception PR done soon and I'll use that.

JasonTheAdams replied:
Finally resolved in de7de24!

felixarntz replied:
My comment here is still relevant I think: What I'm saying is it doesn't make sense to interpret a `WP_Error`'s error code as an integer, since it's meant to be a string.

felixarntz replied:
Per the above, I think we should simply remove this line. There's no reasonable way to maintain the WordPress error code within the exception, unless we come up with a new exception class that can also store a string-shape as error code. But that's overkill IMO.

felixarntz commented on includes/http/WordPress_HTTP_Client.php:
```diff
@@ -0,0 +1,192 @@
+			'method'      => $request->getMethod(),
+			'headers'     => $this->prepare_headers( $request ),
+			'body'        => $this->prepare_body( $request ),
+			'timeout'     => 30,
+			'redirection' => 5,
```
This is really not great - it's not a problem of this PR, but something we need to fix soon in https://github.com/WordPress/php-ai-client: It needs to be possible for a request to specify additional options for how the request should be sent.

I believe we talked about this a few weeks back but decided to defer it. Now that 0.1.0 is out though, it's an important limitation to address. There are many AI calls for which whatever the default timeout of different client implementations is is not sufficient.

JasonTheAdams replied:
Yeah, this one is tricky because the PSR interfaces don't provide a way of including these sorts of parameters. Would you mind making an Issue for this?

felixarntz replied:
I just opened https://github.com/WordPress/php-ai-client/issues/99, would be great to get your thoughts there!

JasonTheAdams replied:
Just opened https://github.com/WordPress/php-ai-client/pull/109 to add the feature!

JasonTheAdams replied:
And now we can resolve this in 4e80b89

felixarntz commented on includes/http/WordPress_HTTP_Client.php:
```diff
@@ -0,0 +1,192 @@
+			'blocking'    => true,
+		);
+
+		// Handle streaming requests if needed.
+		if ( $request->hasHeader( 'X-Stream' ) ) {
+			$args['stream']   = true;
+			$args['filename'] = $request->getHeaderLine( 'X-Stream-Filename' );
+		}
```
This has nothing to do with the type of streaming that we need for AI requests, so I think it would be best to remove this.

WordPress only allows streaming to a file, which does not work when we want to stream HTTP response outputs. I've looked into this for a _long_ time when implementing AI Services, and WordPress simply can't do it.

TL;DR: We will need to use some alternative system to make streaming requests in WordPress, or we need to enhance this in WordPress Core (via https://github.com/WordPress/Requests).

Neither is a particularly great option: The former will require lots of workarounds on our end, the latter will require significant interactions with another project that barely has any active maintainership.

For pragmatic reasons and to not have this block us for a very long time, I would opt for the former. We can e.g.:
* Check if it's a streaming request.
* If so, use Guzzle.
* Either manually duplicate a bunch of WordPress logic to have the necessary hooks run before the request, or try to somehow intercept the WordPress function so that it only does all the request preparation but doesn't send the actual request so that we can handle it.

Anyway, this will require a different issue on https://github.com/WordPress/php-ai-client to figure out.

JasonTheAdams replied:
Oof, yeah, I remember this coming up before. I resolved it here in [581de9e](https://github.com/WordPress/wp-ai-client/pull/2/commits/581de9eec18f839053cfce771797057c45213e05).

Would you mind, @felixarntz, creating an Issue for this in the PHP AI Client since you have your head most wrapped around the problem and options?

felixarntz replied:
I think this technically belongs to this repo, since it's about a WordPress-specific limitation. I opened #11 with more context.

felixarntz commented on includes/HTTP/WordPress_HTTP_Client.php:
```diff
@@ -0,0 +1,192 @@
+		if ( is_array( $headers ) || $headers instanceof \Traversable ) {
+			foreach ( $headers as $name => $value ) {
+				// PSR-7 expects string name and string|array value.
+				$response = $response->withHeader( $name, $value );
```
Will this append in the case of multiple headers of the same name? Just want to make sure this doesn't accidentally overwrite and maintain only the last one.

JasonTheAdams replied:
Since this is using the PSR-7 interface, it's technically up to the concrete response class as to how it handles this. According to the HTTP spec multiple headers with the same name are supported, so it _should_ work that way.

felixarntz replied:
Hmm, now I actually wonder why I asked this question: Looking at the code with fresh eyes, we iterate through `$headers` where `$name` is a key, so in this particular logic there would be no way to pass the same header name multiple times. Not sure what I was thinking about before 😆

felixarntz reviewed (changes requested):
@JasonTheAdams Overall, this all makes sense to me. A bit of feedback regarding the WordPress specific implementation.

Commit:
refactor: renames class

Commit:
refactor: removes stream args

JasonTheAdams committed:
feat: updates to PHP AI Client 0.2.0

JasonTheAdams committed:
feat: implements ClientWithOptionsInterface

JasonTheAdams committed:
refactor: switches to using NetworkException

felixarntz commented on includes/HTTP/WP_AI_Client_Discovery_Strategy.php:
```diff
@@ -0,0 +1,94 @@
+			);
+		}
+
+		// PSR-17 factories - Nyholm's Psr17Factory implements all of them.
+		$psr17_factories = array(
```
This is mostly a question for my own understanding: Why do we need this? Why wouldn't we be able to _only_ inject our WordPress HTTP Client and leave the rest?

JasonTheAdams replied:
So our Client is returned above when the PSR-18 ClientInterface is being asked for. When a PSR-17 factory is asked for, we use the Nyholm one as we don't need to create a custom one for generating PSR Messages and such.

felixarntz commented on includes/http/WP_AI_Client_Discovery_Strategy.php:
```diff
@@ -0,0 +1,94 @@
+			return array(
+				array(
+					'class'     => Psr17Factory::class,
+					'condition' => Psr17Factory::class,
```
What is the `condition` field for? And why is it simply the same as `class`?

JasonTheAdams replied:
It's not. That was added by the AI for some reason. Resolved in 7943ed6

felixarntz commented on includes/HTTP/WP_AI_Client_Discovery_Strategy.php:
```diff
@@ -0,0 +1,94 @@
+	 * @since n.e.x.t
+	 * @return void
+	 */
+	public static function init() {
```
I assume this method is the entry point to all this? As in: Consuming code will need to call this static method to wire up our WordPress HTTP Client with the PHP AI Client SDK, and nothing else, right?

JasonTheAdams replied:
That is correct. We just need something to run this so that the discovery system is aware of this strategy. Seems much simpler than having the implementing system add this to the discovery system itself.

felixarntz commented on includes/http/WordPress_HTTP_Client.php:
```diff
@@ -0,0 +1,241 @@
+		/** Ignoring PHPStan for WordPress-specific array structure. @phpstan-ignore-next-line */
+		$response = \wp_remote_request( $url, $args );
+
+		if ( \is_wp_error( $response ) ) {
+			$message = sprintf(
+				'Network error occurred while sending request to %s: %s',
+				$url,
+				$response->get_error_message()
+			);
```
Minor suggestion: Include the request method here as well.
```suggestion
			$message = sprintf(
				'Network error occurred while sending %s request to %s: %s',
				$request->getMethod(),
				$url,
				$response->get_error_message()
			);
```

JasonTheAdams replied:
Resolved in 7943ed6

felixarntz reviewed (approved):
@JasonTheAdams This looks excellent, great work!

A few questions, and two small things to ideally address before merge.

JasonTheAdams committed:
refactor: cleans up unnecessary conditions code

felixarntz committed:
Merge branch 'trunk' into http-adapter

felixarntz committed:
Use consistent YAML formatting in PHPStan config.

felixarntz committed:
Fix PHPStan errors after updates.

felixarntz committed:
Temp folder rename to fix incorrect casing.

felixarntz committed:
Fix folder name.

felixarntz committed:
Fix PHP warning.

felixarntz committed:
Fix bug in HTTP client discovery implementation.

felixarntz committed:
Wire up HTTP client discovery implementation.

felixarntz commented:
@JasonTheAdams After merging #6 into this, I had to resolve some merge conflicts and minor follow-up fixes for PHPStan stuff.

More importantly though, I want to highlight two improvements:
* I fixed the folder name `includes/http` to be `includes/HTTP` (in line with the namespace).
* I added a line to the single file demo plugin to actually wire up the WordPress HTTP client discovery. See here: https://github.com/WordPress/wp-ai-client/pull/2/commits/a8aaf959781116a29cc9430ff60ebc81c0109569
* There was a bug in the HTTP client discovery implementation that caused a fatal error as soon as I did the above. Fixed here: https://github.com/WordPress/wp-ai-client/pull/2/commits/c50d84b15d4bd30f9873d215cc1990ca948b022e (needs a callable, not an instance)

If you're happy with these changes, please merge 🚀
